test/clienthellotest.c

```diff
@@ -81,50 +74,47 @@ int main(int argc, char *argv[])
         }
 
         len = BIO_get_mem_data(wbio, (char **)&data);
-        dataend = data + len;
+        if (!PACKET_buf_init(&pkt, data, len))
+            goto end;
 
         /* Skip the record header */
-        data += SSL3_RT_HEADER_LENGTH;
+        if (!PACKET_forward(&pkt, SSL3_RT_HEADER_LENGTH))
+            goto end;
+
         /* Skip the handshake message header */
-        data += SSL3_HM_HEADER_LENGTH;
+        if (!PACKET_forward(&pkt, SSL3_HM_HEADER_LENGTH))
+            goto end;
+
         /* Skip client version and random */
-        data += CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE;
-        if (data + SESSION_ID_LEN_LEN > dataend)
+        if (!PACKET_forward(&pkt, CLIENT_VERSION_LEN + SSL3_RANDOM_SIZE))
             goto end;
+
         /* Skip session id */
-        tmplen = *data;
-        data += SESSION_ID_LEN_LEN + tmplen;
-        if (data + CIPHERS_LEN_LEN > dataend)
+        if (!PACKET_get_length_prefixed_1(&pkt, &pkt2))
             goto end;
+
         /* Skip ciphers */
-        tmplen = ((*data) << 8) | *(data + 1);
-        data += CIPHERS_LEN_LEN + tmplen;
-        if (data + COMPRESSION_LEN_LEN > dataend)
+        if (!PACKET_get_length_prefixed_2(&pkt, &pkt2))
             goto end;
+
         /* Skip compression */
-        tmplen = *data;
-        data += COMPRESSION_LEN_LEN + tmplen;
-        if (data + EXTENSIONS_LEN_LEN > dataend)
+        if (!PACKET_get_length_prefixed_1(&pkt, &pkt2))
             goto end;
+
         /* Extensions len */
-        tmplen = ((*data) << 8) | *(data + 1);
-        data += EXTENSIONS_LEN_LEN;
-        if (data + tmplen > dataend)
+        if (!PACKET_as_length_prefixed_2(&pkt, &pkt2))
             goto end;
 
         /* Loop through all extensions */
-        while (tmplen > EXTENSION_TYPE_LEN + EXTENSION_SIZE_LEN) {
-            type = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_TYPE_LEN;
-            size = ((*data) << 8) | *(data + 1);
-            data += EXTENSION_SIZE_LEN;
-            if (data + size > dataend)
+        while (PACKET_remaining(&pkt2)) {
+
+            if (!PACKET_get_net_2(&pkt2, &type) ||
+                !PACKET_get_length_prefixed_2(&pkt2, &pkt))
```

Comment:
It might. Do you realise *just* how far down the road I've come from my initial _"oh, DTLS1_BAD_VER broke again..."_ ? :)